<?xml version="1.0"?>
<launch>
  <!-- Arguments -->
  <arg name="model" default="$(env TURTLEBOT3_MODEL)" doc="model type [burger, waffle, waffle_pi]"/>
  <arg name="x_pos" default="0.0"/>
  <arg name="y_pos" default="-3.0"/>
  <arg name="z_pos" default="0.0"/>
  <arg name="gui" default="true"/>

  <!-- Launch Gazebo with maze world -->
  <include file="$(find gazebo_ros)/launch/empty_world.launch">
    <arg name="world_name" value="$(find maze_runner)/worlds/simple_box_maze.world"/>
    <arg name="paused" value="false"/>
    <arg name="use_sim_time" value="true"/>
    <arg name="gui" value="$(arg gui)"/>
    <arg name="headless" value="false"/>
    <arg name="debug" value="false"/>
  </include>

  <!-- Spawn TurtleBot3 -->
  <param name="robot_description" command="$(find xacro)/xacro --inorder $(find turtlebot3_description)/urdf/turtlebot3_$(arg model).urdf.xacro" />
  <node name="spawn_urdf" pkg="gazebo_ros" type="spawn_model" args="-urdf -model turtlebot3_$(arg model) -x $(arg x_pos) -y $(arg y_pos) -z $(arg z_pos) -param robot_description" />

  <!-- Robot state publisher -->
  <node pkg="robot_state_publisher" type="robot_state_publisher" name="robot_state_publisher">
    <param name="publish_frequency" type="double" value="30.0" />
    <param name="tf_prefix" value="" />
  </node>

  <!-- Map server (start with empty map for SLAM) -->
  <node name="map_server" pkg="map_server" type="map_server" args="$(find turtlebot3_navigation)/maps/map.yaml"/>

  <!-- AMCL for localization -->
  <include file="$(find turtlebot3_navigation)/launch/amcl.launch"/>

  <!-- Move base for autonomous navigation with smooth parameters -->
  <include file="$(find maze_runner)/launch/smooth_move_base.launch">
    <arg name="model" value="$(arg model)" />
  </include>

  <!-- SLAM (Gmapping) -->
  <include file="$(find turtlebot3_slam)/launch/turtlebot3_gmapping.launch"/>

  <!-- Path recorder node -->
  <node name="path_recorder" pkg="maze_runner" type="path_recorder.py" output="screen"/>

  <!-- Path retractor node -->
  <node name="path_retractor" pkg="maze_runner" type="path_retractor.py" output="screen">
    <param name="simplify_path" value="true"/>
    <param name="goal_timeout" value="30.0"/>
    <param name="max_consecutive_failures" value="2"/>
    <param name="xy_goal_tolerance" value="0.3"/>
    <param name="min_keypoint_distance" value="1.0"/>
    <param name="angle_threshold" value="30.0"/>
    <param name="max_keypoints" value="15"/>
  </node>

  <!-- Teleop keyboard control in external terminal -->
  <node name="turtlebot3_teleop_keyboard" pkg="turtlebot3_teleop" type="turtlebot3_teleop_key" 
        output="screen" launch-prefix="xterm -title 'TurtleBot3 Teleop Controls - Use WASD to move!' -geometry 80x20 -e"/>

  <!-- RViz for visualization -->
  <node name="rviz" pkg="rviz" type="rviz" args="-d $(find maze_runner)/param/maze_exploration.rviz" required="false"/>
</launch>
